discord/voice_client.py

```diff
@@ -293,23 +393,180 @@ def is_connected(self):
+        The returned data is stereo 48kHz audio in raw PCM. You will have to
+        add your own headers if you want to save it as a `.wav` file.
+        """
+        buffer = []
+        last_packet = 0
+        user_id = user.id
+
+        def listener(vc, vp):
+            nonlocal buffer, user_id, last_packet
+            if user_id == (None if vp.user is None else vp.user.id):
+                last_packet = time.time()
+                heapq.heappush(buffer, (vp.timestamp, vp))
+
+        self._pcm_listeners.append(listener)
+
+        while time.time() - last_packet < timeout or last_packet == 0:
+            yield from asyncio.sleep(0.1)  # Continue to fill up the buffer.
+        self._pcm_listeners.remove(listener)
+
+        buffer.sort()
+        data = bytearray()
+        current_timestamp = 0
+        for key, value in buffer:
+            if current_timestamp == 0:
+                delta = 0
+            else:
+                delta = (key - current_timestamp) - 960
+                delta = max(delta, 0)  # Just to be safe
+
+            padding = bytearray([0] * delta * 4)
+            data += padding
+            data += value.pcm
+
+            current_timestamp = key
+
+        return data
+
+    @asyncio.coroutine
+    def pipe_voice_into_file(self, user, file_object, buffer_size=0.2):
+        """|coro|
 
-        # Encrypt and return the data
-        return header + box.encrypt(bytes(data), bytes(nonce)).ciphertext
+        Constantly receive data from a user and pipe it into a file-like
+        object. Packets will be reordered and any silence or missed packets
+        will be accounted for.
+
+        Parameters
+        -----------
+        user: :class:`abc.User`
+            The user to listen for audio from.
+        file_object: File-like
+            The file-like object to pipe the voice data into. This object just
+            has to have a `.write` method that takes bytes.
+        buffer_size: Optional[float]
+            The number of seconds to bufffer the audio for before writing to
+            the `file_object`. The larger this value, the less likely you are
+            to miss a packet, but the longer you have to wait before the data
+            is in the `file_object`.
+
+        Note
+        -----------
+        This function, although a co-routine, is not blocking.
+
+        Note
+        -----------
+        The returned data is stereo 48kHz audio in raw PCM. You will have to
+        add your own headers if you want to save it as a `.wav` file.
+        """
+
+        # Convert buffer_size from seconds to packets.
+        buffer_size = int(round((buffer_size * 1000) / 20))
+
+        SAMPLE_RATE = 48000
+        SAMPLE_SIZE = 2
+
+        last_packet = 0
+        user_id = user.id
+        voice_buffer = []
+        last_speaking = 0
+
+        def listener(vc, vp):
+            nonlocal voice_buffer, user_id, last_packet
+            if user_id == (None if vp.user is None else vp.user.id):
```

Comment:
u should probably drop packets that are too old (ie part of segments that have already been decoded) - this handles late packets.

Reply:
This is handled by the piping function. The raw callbacks don't yet have a jitter buffer on them, but I'm planning on adding one.

Reply:
(Line 510)